lib/_http_parser.js

```diff
@@ -0,0 +1,895 @@
+var RES_HTTP_VER_BYTES = REQ_HTTP_VER_BYTES.slice(1);
+REQ_HTTP_VER_BYTES.reverse();
+
+var STATE_REQ_LINE = 0;
+var STATE_STATUS_LINE = 1;
+var STATE_HEADER = 2;
+var STATE_BODY_LITERAL = 3;
+var STATE_BODY_EOF = 4;
+var STATE_BODY_CHUNKED_SIZE = 5;
+var STATE_BODY_CHUNKED_SIZE_IGNORE = 6;
+var STATE_BODY_CHUNKED_BYTES = 7;
+var STATE_BODY_CHUNKED_BYTES_LF = 8;
+var STATE_COMPLETE = 9;
+
+var FLAG_CHUNKED = 1 << 0;
+var FLAG_CONNECTION_KEEP_ALIVE = 1 << 1;
+var FLAG_CONNECTION_CLOSE = 1 << 2;
+var FLAG_CONNECTION_UPGRADE = 1 << 3;
+var FLAG_TRAILING = 1 << 4;
+var FLAG_UPGRADE = 1 << 5;
+var FLAG_SKIPBODY = 1 << 6;
+var FLAG_SHOULD_KEEP_ALIVE = 1 << 7;
+var FLAG_CONNECT_METHOD = 1 << 8;
+var FLAG_ANY_UPGRADE = FLAG_UPGRADE | FLAG_CONNECTION_UPGRADE;
+
+function HTTPParser(type) {
+  if (type === RESPONSE) {
+    this.type = type;
+    this._state = STATE_STATUS_LINE;
+  }
+  this.headers = [];
+}
+HTTPParser.prototype.type = REQUEST;
+HTTPParser.prototype._state = STATE_REQ_LINE;
+HTTPParser.prototype._err = null;
+HTTPParser.prototype._flags = 0;
+HTTPParser.prototype._contentLen = null;
+HTTPParser.prototype._nbytes = null;
+HTTPParser.prototype._nhdrbytes = 0;
+HTTPParser.prototype._nhdrpairs = 0;
+HTTPParser.prototype._buf = '';
+HTTPParser.prototype.httpMajor = 1;
+HTTPParser.prototype.httpMinor = null;
+HTTPParser.prototype.maxHeaderPairs = 2000;
+HTTPParser.prototype.method = null;
+HTTPParser.prototype.url = null;
+HTTPParser.prototype.statusCode = null;
+HTTPParser.prototype.statusText = null;
+
+HTTPParser.prototype.onHeaders = null;
+HTTPParser.prototype.onBody = null;
+HTTPParser.prototype.onComplete = null;
+
+HTTPParser.prototype.onIncoming = null;
+HTTPParser.prototype.incoming = null;
+HTTPParser.prototype.socket = null;
+
+HTTPParser.prototype.reinitialize = function(type) {
+  this.execute = this._executeStartLine;
+  this.type = type;
+  if (type === REQUEST)
+    this._state = STATE_REQ_LINE;
+  else
+    this._state = STATE_STATUS_LINE;
+
+  this._err = null;
+  this._flags = 0;
+  this._contentLen = null;
+  this._nbytes = null;
+  this._nhdrbytes = 0;
+  this._nhdrpairs = 0;
+  this._buf = '';
+
+  // Common properties
+  this.headers = [];
+  this.httpMajor = 1;
+  this.httpMinor = null;
+  this.maxHeaderPairs = 2000;
+
+  // Request properties
+  this.method = null;
+  this.url = null;
+
+  // Response properties
+  this.statusCode = null;
+  this.statusText = null;
+};
+HTTPParser.prototype.finish = function() {
+  var state = this._state;
+  if (state === STATE_BODY_EOF) {
+    this.execute = this._executeBodyIgnore;
+    this._state = STATE_COMPLETE;
+    this.onComplete && this.onComplete();
+  } else if (this.execute !== this._executeError &&
+             state !== STATE_REQ_LINE &&
+             state !== STATE_STATUS_LINE &&
+             state !== STATE_COMPLETE) {
+    return this._setError('Invalid EOF state');
+  }
+};
+HTTPParser.prototype.close = function() {};
+HTTPParser.prototype.pause = function() {};
```

Comment:
Yeah I noted this in my initial write-up, since AFAICT the `pause()`/`resume()` are not being used properly, they can probably be safely removed unless my understanding of how the `pause()` and `resume()` are currently being used are not correct.